crates/apollo-compiler/src/database/inputs.rs

```diff
@@ -9,19 +9,63 @@ use std::sync::Arc;
 #[error("Unknown file ID")]
 struct UnknownFileError;
 
+/// Source text structure with a precomputed index from byte offsets to character offsets and to
+/// line/column numbers.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MappedSource {
+    ariadne: AriadneSource,
+    map: Vec<u32>,
+}
+
+impl MappedSource {
+    fn new(input: &str) -> Self {
+        let ariadne = AriadneSource::from(input);
+
+        let mut map = vec![0; input.len() + 1];
+        let mut char_index = 0;
+        for (byte_index, _) in input.char_indices() {
+            map[byte_index] = char_index;
+            char_index += 1;
+        }
+
+        // Support 1 past the end of the string, for use in exclusive ranges.
+        map[input.len()] = char_index;
+
+        Self { ariadne, map }
+    }
+
+    pub(crate) fn map_index(&self, byte_index: usize) -> usize {
+        self.map[byte_index] as usize
+    }
+}
+
 /// A Cache implementation for `ariadne` diagnostics.
 ///
 /// Use [`InputDatabase::source_cache`] to construct one.
 #[derive(Clone, PartialEq, Eq)]
 pub struct SourceCache {
-    sources: HashMap<FileId, Arc<AriadneSource>>,
+    sources: HashMap<FileId, Arc<MappedSource>>,
     paths: HashMap<FileId, PathBuf>,
 }
+
+impl SourceCache {
+    pub(crate) fn get_source(&self, id: FileId) -> Option<&'_ MappedSource> {
+        self.sources.get(&id).map(|arc| arc.as_ref())
+    }
+
+    pub fn get_line_column(&self, id: FileId, index: usize) -> Option<(usize, usize)> {
```

Comment:
should line/column be its own struct or be part of another struct? feels a bit strange api-wise with a tuple. i imagine there is going to be loads of use cases for this functionality

Reply:
+1, `GraphQLLocation` is being added anyway so using it here instead of a tuple makes sens.

Reply:
Haha good thought, I added graphqllocation after adding this method, totally agree.